src/linear/updater_gpu_coordinate.cu

```diff
@@ -235,29 +145,100 @@ class GPUCoordinateUpdater : public LinearUpdater {  // NOLINT
+      UpdateResidual(dw, group_idx, model->param.num_output_group, fidx);
     }
   }
 
+  // This needs to be public because of the __device__ lambda.
+  GradientPair GetBiasGradient(int group_idx, int num_group) {
+    dh::safe_cuda(cudaSetDevice(device_));
+    auto counting = thrust::make_counting_iterator(0ull);
+    auto f = [=] __device__(size_t idx) {
+      return idx * num_group + group_idx;
+    };  // NOLINT
+    thrust::transform_iterator<decltype(f), decltype(counting), size_t> skip(
+        counting, f);
+    auto perm = thrust::make_permutation_iterator(gpair_.data(), skip);
+
+    return dh::SumReduction(temp_, perm, num_row_);
+  }
+
+  // This needs to be public because of the __device__ lambda.
+  void UpdateBiasResidual(float dbias, int group_idx, int num_groups) {
+    if (dbias == 0.0f) return;
+    auto d_gpair = gpair_;
+    dh::LaunchN(device_, num_row_, [=] __device__(size_t idx) {
+      auto &g = d_gpair[idx * num_groups + group_idx];
+      g += GradientPair(g.GetHess() * dbias, 0);
+    });
+  }
+
+  // This needs to be public because of the __device__ lambda.
+  GradientPair GetGradient(int group_idx, int num_group, int fidx) {
+    dh::safe_cuda(cudaSetDevice(device_));
+    common::Span<xgboost::Entry> d_col = data_.subspan(row_ptr_[fidx]);
+    size_t col_size = row_ptr_[fidx + 1] - row_ptr_[fidx];
+    common::Span<GradientPair> d_gpair = gpair_;
+    auto counting = thrust::make_counting_iterator(0ull);
+    auto f = [=] __device__(size_t idx) {
+      auto entry = d_col[idx];
+      auto g = d_gpair[entry.index * num_group + group_idx];
+      return GradientPair(g.GetGrad() * entry.fvalue,
+                          g.GetHess() * entry.fvalue * entry.fvalue);
+    };  // NOLINT
+    thrust::transform_iterator<decltype(f), decltype(counting), GradientPair>
+        multiply_iterator(counting, f);
+    return dh::SumReduction(temp_, multiply_iterator, col_size);
+  }
+
+  // This needs to be public because of the __device__ lambda.
+  void UpdateResidual(float dw, int group_idx, int num_groups, int fidx) {
+    common::Span<GradientPair> d_gpair = gpair_;
+    common::Span<Entry> d_col = data_.subspan(row_ptr_[fidx]);
+    size_t col_size = row_ptr_[fidx + 1] - row_ptr_[fidx];
+    dh::LaunchN(device_, col_size, [=] __device__(size_t idx) {
+      auto entry = d_col[idx];
+      auto &g = d_gpair[entry.index * num_groups + group_idx];
+      g += GradientPair(g.GetHess() * dw * entry.fvalue, 0);
+    });
+  }
+
  private:
+  bool IsEmpty() {
+    return num_row_ == 0;
+  }
+
+  void UpdateGpair(const std::vector<GradientPair> &host_gpair,
+                   const gbm::GBLinearModelParam &model_param) {
+    dh::safe_cuda(cudaMemcpyAsync(
+        gpair_.data(),
+        host_gpair.data(),
+        gpair_.size() * sizeof(GradientPair), cudaMemcpyHostToDevice));
+  }
+
   // training parameter
   LinearTrainParam tparam_;
   CoordinateParam coord_param_;
-  int device_{};
+  int device_{-1};
```

Comment:
We should use the `generic_parameter` in the future.

Reply:
Done.